// muse-ui/lib/DateInput/DateInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _TextField = require('../TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _DatePicker = require('../DatePicker');

var _DatePicker2 = _interopRequireDefault(_DatePicker);

var _Container = require('./Container');

var _Container2 = _interopRequireDefault(_Container);

var _TimePicker = require('../TimePicker');

var _TimePicker2 = _interopRequireDefault(_TimePicker);

var _dayjs = require('dayjs');

var _dayjs2 = _interopRequireDefault(_dayjs);

var _Button = require('../Button/Button');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_FORMAT = {
  date: 'YYYY-MM-DD',
  time: 'hh:mm',
  year: 'YYYY',
  month: 'YYYY-MM',
  dateTime: 'YYYY-MM-DD hh:mm'
};

var PickerProps = (0, _extends3.default)({}, _TimePicker2.default.props, _DatePicker2.default.props);

delete PickerProps.date;
delete PickerProps.time;
delete PickerProps.type;
delete PickerProps.format;
exports.default = {
  name: 'mu-date-input',
  model: {
    prop: 'value',
    event: 'input'
  },
  props: (0, _extends3.default)({
    container: {
      type: String,
      default: 'popover', validator: function validator(val) {
        return val && ['dialog', 'popover', 'bottomSheet'].indexOf(val) !== -1;
      }
    },
    type: {
      type: String,
      default: 'date' },
    format: {
      type: String
    },
    rangeSeparator: {
      type: String,
      default: '—'
    },
    actions: Boolean,
    clockType: _TimePicker2.default.props.format,
    okLabel: {
      type: String,
      default: '确定'
    },
    cancelLabel: {
      type: String,
      default: '取消'
    },
    value: {},
    valueFormat: String
  }, _TextField2.default.props, PickerProps),
  data: function data() {
    return {
      open: false,
      date: this.value ? (0, _dayjs2.default)(this.value).toDate() : undefined,
      view: this.type === 'time' ? 'time' : this.type === 'dateRange' ? 'dateRange' : 'date'
    };
  },

  methods: {
    changeValue: function changeValue() {
      this.closePicker();
      var value = this.valueFormat ? (0, _dayjs2.default)(this.date).format(this.valueFormat) : this.date;
      this.$emit('change', value);
      this.$emit('input', value);
    },
    closePicker: function closePicker() {
      this.open = false;
      this.view = this.type === 'time' ? 'time' : this.type === 'dateRange' ? 'dateRange' : 'date';
    },
    handleDateChange: function handleDateChange(date) {
      this.date = date;
      if (this.type === 'dateTime') {
        this.view = 'time';
        return;
      }

      if (!this.actions) this.changeValue();
    },
    handleTimeChange: function handleTimeChange(date, mode, finished) {
      this.date = date;
      if (!finished || mode !== 'minute') return;
      if (!this.actions) this.changeValue();
    },
    generateTextFieldProps: function generateTextFieldProps() {
      return this.generateProps(_TextField2.default.props);
    },
    generateDatePickerProps: function generateDatePickerProps() {
      return this.generateProps(_DatePicker2.default.props);
    },
    generateTimePickerProps: function generateTimePickerProps() {
      return this.generateProps(_TimePicker2.default.props);
    },
    generateProps: function generateProps(props) {
      var _this = this;

      var obj = {};
      (0, _keys2.default)(props).forEach(function (key) {
        obj[key] = _this[key];
      });
      return obj;
    },
    createActions: function createActions(h, name) {
      if (!this.actions) return;
      return h('div', {
        staticClass: 'mu-' + name + '-actions'
      }, [h(_Button2.default, {
        props: {
          flat: true,
          color: 'primary'
        },
        on: {
          click: this.closePicker
        }
      }, this.cancelLabel), h(_Button2.default, {
        props: {
          flat: true,
          color: 'primary'
        },
        on: {
          click: this.changeValue
        }
      }, this.okLabel)]);
    },
    createPicker: function createPicker(h) {
      return h(_Container2.default, {
        props: {
          container: this.container,
          open: this.open,
          trigger: this.$el
        },
        on: {
          close: this.closePicker
        }
      }, [this.view === 'time' ? h(_TimePicker2.default, {
        props: (0, _extends3.default)({}, this.generateTimePickerProps(), {
          time: this.date,
          noDisplay: false,
          format: this.clockType
        }),
        on: {
          change: this.handleTimeChange
        },
        style: {
          width: this.container === 'bottomSheet' ? 'auto' : ''
        }
      }, [this.createActions(h, 'timepicker')]) : h(_DatePicker2.default, {
        props: (0, _extends3.default)({}, this.generateDatePickerProps(), {
          type: this.type === 'month' ? 'month' : this.type === 'year' ? 'year' : 'date',
          date: this.date
        }),
        on: {
          change: this.handleDateChange
        },
        style: {
          width: this.container === 'bottomSheet' ? 'auto' : ''
        }
      }, [this.createActions(h, 'datepicker')])]);
    }
  },
  render: function render(h) {
    var _this2 = this;

    var dateStr = this.value ? (0, _dayjs2.default)(this.value).format(this.format ? this.format : DEFAULT_FORMAT[this.type]) : '';
    var listeners = this.$listeners;
    delete listeners.input;
    delete listeners.change;
    return h(_TextField2.default, {
      props: (0, _extends3.default)({}, this.generateTextFieldProps(), {
        value: dateStr
      }),
      attrs: (0, _extends3.default)({
        readonly: true
      }, this.$attrs),
      on: (0, _extends3.default)({}, listeners, {
        click: function click(e) {
          _this2.open = true;
          _this2.$emit('click', e);
        }
      })
    }, [this.createPicker(h)]);
  },

  watch: {
    value: function value(val) {
      this.date = val ? (0, _dayjs2.default)(val).toDate() : undefined;
    }
  }
};
